### ISSUE TITLE
Document for functions to be implemented

### ISSUE THREAD

**Issue by Gaoooooo** [Hedgehog-Computing/math#6]
Attached is the document for functions to be implemented: 

They are also listed here but not fully up do date: 

Mathjs functions: 

factorial (of a number and an array element-wise)
det (determinant of a matrix)
add(adds a constant to an array/matrix)
lusolve(finds one solution of linear equation system)
lsolve (find ONE solution of linear equation systems by forward substitution. Has to be lower triangular matrix)
lsolveAll -> [rename as lsolve_all] (find ALL solutions of linear equation system)
usolve  (find one solution by back substitution)
usolveAll -> [rename as usolve_all] (find all solutions of linear equation system)
slu (sparse matrix LU decomposition)
abs (returns the absolute value of a number or matrix element-wise - done in base?
cbrt (cubic root —-> look to implement powers/exponents? Maybe use math.pow
ceil  (rounds number to next largest integer)
floor (rounds number down, to next smallest integer)
dotDivide → [rename as dot_divide(x, y)] (divides 2 matrices element wise)
dotMultiply →[rename as dot_multiply] (multiplies 2 matrices element wise)
exp (exponent, again look to implement powers, roots) use math.pow
gcd(a,b) (greatest common divisor of a, b)
lcm(a,b) (least common multiple of a,b)
log(x,[base]) (log of x given a base)
mod(x, y) (x modulo y)

diag(X) (creates a diagonal matrix or retrieve the diagonal of a matrix when x is a vector a matrix with vector x on the diagonal will be returned)
eigs(x, [prec]) (eigenvalues of a matrix with a precision amount)
flatten(x) (makes a multidimensional array 1-d)
identity(n) (creates 2-d identity matrix with nxn size) 
inv(x) (given a square matrix, finds the inverse)
ones(m,n,p, … ) (gives an array of m x n x p x … array of 1’s with dimensions m,n,p,...
reshape (x, sizes) (reshapes a multidimensional array x into an array with dimensions of sizes which is an array such as [2,2,2] or [2,3])
resize(x, size, [defaultValue]) (resize a matrix x to be of size ‘size’ and any new elements will have value ‘defaultValue’)
rotationMatrix(theta in radians) (creates a 2d counter clockwise rotation 2x2 rotation matrix for a given angle in radians 
multinomial(a) ( computes number of ways of picking a1, a2, … ai unordered outcomes from n possibilities, where a is array of integers>=0)
sum(a,b,c,...) (sum of a,b,c,... list or if given a matrix, some of all elements)
max(a,b,c….) (max element of a,b,c… or given a matrix max element in matrix)
min(a,b,c….) (min element of a,b,c… or given a matrix min element in matrix)
mean(a,b,c,....) or mean(A) (mean of a time series or mean of a matrix element wise)
median(a,b,c….) (median of a list or matrix element wise)
mode(a,b,c…) (mode of a list or matrix element wise)
std(a,b,c,....) (standard deviation of a list or matrix element wise)
variance (a,b,c,....)(variance of a list or matrix element wise)

Matlab: 

Preprocessing data:
any_missing  (determine if any array element is missing)
is_missing (find missing values)
rmmissing (remove missing entries)
fillmissing (fill missing values)
missing  (create missing values)
standardize_missing (insert standard missing values)
is_outlier  (find outliers in data)
fill_outliers (detect and replace outliers in data)
rm_outliers (detect and remove outliers in data)
is_local_min  (find local minima)
is_local_max  (find local maxima)
smooth_data  (smooth noisy data)
trend_decomp (find trends in data)
normalize (normalize data)
Grouping and binning data?

Descriptive statistics:
mink  (find k smallest elements of array)
maxk  (find k largest elements of array)
bounds (find min and max of array)
mean/median/mode/std/var (covered in mathjs
iqr  (interquartile range of data set)
quantile (quantiles of data set) 
prctile (percentiles of data set) (-> rename as percentile?)
rms  (root mean square value)
correcoef ( correlation coefficient)
cov (covariance) 
xcorr  (cross correlation)
xcov  (cross-covariance)

Trigonometry: base code issue for now

Test matrices - 
hardamard (hadamard matrix)
hankel (hankel matrix)
hilb (hilbert matrix)
invhilb (inverse of hilbert matrix)
magic (magic square)
pascal (pascal matrix)
toeplitz (toeplitz matrix)
vander (vander matrix)
wilkinson (wilkinsons eigenvalue test matrix)

Linear Algebra:
(not in matlab - convolution)
svd  (singular value decomp)
lsq (least square solution to linear equation)
pinv  (moore-penrose psuedoinverse)
lscov (least square solution in presence of knoffwn covariance)
eig (eigenvalues and eigenvectors in other contexts than mathjs)
eigs (subset of eigenvalues and eigenvectors)
schur (schur decomposition of a matrix)
hess (hessenberg form of a matrix)
qz (qz factorization for generalized eigenvalues)
funm (evaluate general matrix function)
kron  (kronecker tensor product)
cross  (cross product)
dot  (dot product)
is_diag (boolean is diagonal matrix or not)
is_tril  (boolean is lower triangular or not) && is_triu (boolean is upper triangular or not)
condeig  (condition number with respect to eigenvalue)

Machine Learning: (https://github.com/Hedgehog-Computing/ml)

Linear regression (iterative method and numeric method)
Libsvm
Lasso
Logistic regression 
Wrapper for tensorflow.js

Xgboost: the SOTA classical solver for machine learning
API and dataset:
Rapid API: using fetch or axios to fetch data “synchronously” 
function rapidAPI(api_name, api_key) {... return result}
Kaggle.com

Graph and network algorithms 
*data structures

(basic matrix octave) https://octave.org/doc/v4.2.0/Basic-Matrix-Functions.html#Basic-Matrix-Functions

(matrix factorizations plus more octave)
https://octave.org/doc/v4.2.0/Matrix-Factorizations.html#Matrix-Factorizations

(matlab mathematics functions)
https://www.mathworks.com/help/matlab/referencelist.html?type=function&listtype=cat&category=mathematics&blocktype=all&capability=&s_tid=CRUX_lftnav

(matlab all functions)
https://www.mathworks.com/help/matlab/referencelist.html?type=function&listtype=cat&category=index&blocktype=all&capability=&s_tid=CRUX_lftnav

Sparse matrices (Matlab functions) - speye (sparse identity), sprand (sparse uniformly distributed random matrix), sprandn (sparse normally distributed random matrix), sprandsym (sparse symmetric random matrix), sparse (create sparse matrix) issparse, nnz (number of nonzero elements), spfun (apply function to nonzero elements), sponges (replace nonzero elements with ones), find (find indices and values of nonzero elements), Reordering algorithm…. Iterative methods and preconditioners……eigenvalues and singular values…… structural analysis….. LATER

Diff eq/calculus - no symbolic somewhere else, not nerdamer though (last thing) (check other implementations of newton/diff eq methods open source)

Fourier analysis/filtering - yes, later

Iterative methods - newtons method, golden etc? In javascript? Not yet and somewhere else
